### PR TITLE
Many security updates

### PR DESCRIPTION
**Pull request by ayush-goyal**
- Add 404 and 500 error pages rather than stack traces in prod
Screen Shot 2020-12-27 at 3 45 39 PM
Screen Shot 2020-12-27 at 3 47 06 PM

-----

- Add password complexity/length requirement
- At first, I only added password checking when signing up, but in a later commit I added it in when changing/resetting your password
Screen Shot 2020-12-27 at 3 46 09 PM

-----

- Update all dependencies to latest - some of them add security updates that were needed
- I had to add in new typings and updated the mongo db options because of this

-----

- Switched bugsnag error handling over to sentry since this is what we will be using going forward

-----

- Added CSRF protection for all forms (ex. when changing password, and for all admin endpoints)
- Added in express Helmet for protection
- Refactored express post parser

-----

- Added rate limiting to all endpoints, and created utility functions for different rate limiting options per endpont (in middleware.ts)

### CODE REVIEW

**Comment by ehsanmasdar** [comment]
This is awesome!

> Added rate limiting to all endpoints, and created utility functions for different rate limiting options per endpont (in middleware.ts)
Might be worth adding recaptcha on some endpoints to reduce risk of brute force/spam even further.

**Comment by evan10s** [comment]
Same as Ehsan, these are all excellent improvements to make!  I will look at the code shortly - I agree about recaptcha and have a couple thoughts about password complexity requirements as well

**Comment by ayush-goyal** [comment]
@ehsanmasdar @evan10s I added in recaptcha for certain routes. See [login.dev.hack.gt](login.dev.hack.gt)
Screen Shot 2020-12-28 at 12 03 46 AM

**Review by evan10s — CHANGES_REQUESTED** [review]

File: src/auth/strategies/Local.ts [hunk 1]
```diff
@@ -20,6 +21,13 @@ const pbkdf2Async = async (password: string | Buffer, salt: string | Buffer, rou
     return util.promisify(crypto.pbkdf2).call(null, password, salt, rounds, 128, "sha256");
 };
 
+const passwordSchema = new passwordValidator();
+passwordSchema
```
[the remark this answers is not in the document]
Comment (evan10s) [reply src/auth/strategies/Local.ts:25]: I also lose all the data I entered if I try to submit without completing recaptcha.

**Comment by evan10s** [comment]
Also, when I sign in at login.dev.hack.gt, I get taken to https://login.dev.hack.gt/favicon.ico instead of the 404 page.

**Comment by ayush-goyal** [comment]
> Also, when I sign in at login.dev.hack.gt, I get taken to https://login.dev.hack.gt/favicon.ico instead of the profile page.
Can you explain more? When I sign in I go to the profile page

**Comment by evan10s** [comment]
As crazy as this sounds, it only seems to happen if I click on the icon part of the Next button or if I complete the recaptcha, then click back to the password field and hit Enter (in Chrome on Windows).  It still ends up logging me in though, and both of those seem to be consistent causes

**Comment by ayush-goyal** [comment]
I'm not sure what the problem is and I can't replicate it. Can you try a hard reload and clear your cache?

**Comment by evan10s** [comment]
Yeah, it's still happening in an incognito window.

**Review by evan10s — APPROVED** [review]

**Pull request merged**